### PR TITLE
`ColorPalette`: partial support of `color-mix()` CSS colors

### PR DESCRIPTION
**Pull request by Rishit30G**
## Why?
Users were facing issue with `ColorGradientsControl` showing incorrect labels using `color-mix` 

### What ? 

Related to https://github.com/WordPress/gutenberg/issues/64195

## How?
Solved this issue by adding `const currentValueIsColorMix = /^color-mix\(.*\)$/.test( currentValue );` to check for color-mix and using it logically in `normalizedCurrentValue` instead of converting `color-mix` value to hex which was giving the same label as only the transparency was changing but the color was same i.e `#000000`

## Testing Instructions
Copy the [color palette](https://github.com/WordPress/gutenberg/issues/64195) snippet in this issue to the theme.json.
Go to block editor (e.g. edit a page)
Select a block and try to change its color.
Try switching between the new colors and the selected label will always show 10% of Current Text.

## Screenshots or screencast 

https://github.com/user-attachments/assets/356e8015-316d-4660-9f8b-e2550926fe19

### CODE REVIEW

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [inline packages/components/src/color-palette/utils.ts]: I'd apply a less strict regex here for better matching:

```suggestion
	const currentValueIsColorMix = /color-mix\(/.test( currentValue );
```

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [inline packages/components/src/color-palette/utils.ts:1]: I'm not sure that we can fully support `color-mix` CSS colors at the moment.

I took a look at the component, and noticed that this PR is missing changes in the `normalizeColorValue` function at the bottom of this file. The `normalizeColorValue` function takes care of converting the chosen color option to a CSS color that is passed to the `ColorPicker`.

I applied the same logic in that function too:

```diff
diff --git a/packages/components/src/color-palette/utils.ts b/packages/components/src/color-palette/utils.ts
index 00438db528..43104d2084 100644
--- a/packages/components/src/color-palette/utils.ts
+++ b/packages/components/src/color-palette/utils.ts
@@ -82,9 +82,12 @@ export const normalizeColorValue = (
 	value: string | undefined,
 	element: HTMLElement | null
 ) => {
-	const currentValueIsCssVariable = /^var\(/.test( value ?? '' );
+	const valueIsCssVariable = /var\(/.test( value ?? '' );
+	const valueIsColorMix = /color-mix\(/.test( value ?? '' );
 
-	if ( ! currentValueIsCssVariable || element === null ) {
+	const valueIsSimpleColor = ! valueIsCssVariable && ! valueIsColorMix;
+
+	if ( valueIsSimpleColor || element === null ) {
 		return value;
 	}
 
```

But the computed styles that the browser returns from applying a `color-mix()` color to an element [use the CSS `color()` function](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color), which is not a valid input to `colord` (the library that we use to manipulate and normalize color strings).

And therefore, when opening the `ColorPicker`, the color initially chosen in the picker is not the correct color, but black (which is the default color that `colord` outputs when the string that it parses is not valid).


https://github.com/user-attachments/assets/b7a139b3-acba-4ffb-8724-6e9e8a3f99d5

**Review by ciampo — COMMENTED** [review]
Thank you for working on this!

Unfortunately, as explained in an inline comment, we're currently unable to fully support `color-mix` because of lack of support in parsing newer CSS color syntaxes from an internal dependency.

Having said that, we may also argue that this PR still improves the component with respect to what's on `trunk`, and therefore we may want to still merge such improvements.

@WordPress/gutenberg-components any thoughts?

If we were to continue work on this PR, it would be great also to add unit tests.

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [reply packages/components/src/color-palette/utils.ts]: And potentially consider removing the "start of line" caret (`^`) from `currentValueIsCssVariable`  too

File: packages/components/src/color-palette/utils.ts
Comment (mirka) [reply packages/components/src/color-palette/utils.ts:1]: > But the computed styles that the browser returns from applying a `color-mix()` color to an element [use the CSS `color()` function](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color), which is not a valid input to `colord` (the library that we use to manipulate and normalize color strings).

What are the repro steps for this by the way? I'm just trying it in the browser and not on this PR branch, but when I try it I get a string for the respective color space. So if I `getComputedStyle()` on an element with `background-color: color-mix(in oklab, #a71e14 75%, white)` for example, I will get `'oklab(0.603981 0.113291 0.0640898)'`.

If we can do that, we could possibly add support for it if it's a color space that has a [`colord` plugin](https://github.com/omgovich/colord#plugins). Maybe not in this PR though.

And yes I'm all for merging changes as long as it's incrementally better than trunk 👍

Comment (ciampo) [reply packages/components/src/color-palette/utils.ts:1]: > What are the repro steps for this by the way?

For example, in latest chrome, if I apply `background-color: color-mix(in srgb, indigo, white)` to an element and then query its computed `background-color` style, I get `color(srgb 0.647059 0.5 0.754902)`.

> If we can do that, we could possibly add support for it if it's a color space that has a [`colord` plugin](https://github.com/omgovich/colord#plugins). Maybe not in this PR though.

I had looked into it, and didn't spot any plugin for the `color()` syntax — and not even for newer color spaces like `oklab` and `oklch`

I guess it's up to the browser to decide the format of the computed style string, and since `color-mix` can target any color space, to support fully `color-mix` we'd need to be able to virtually support _any_ color syntax.

Also, looking at `colord` repo, I think the project is not being maintained anymore 😱  Given that CSS continues to evolve, I think that the `colord`-based approach is showing its limitations. We should ideally find a browser-based way (either through CSS or native JS) that will guarantee us that whatever CSS color was passed to the component, we're able to parse it, show it, and render a working color picker.

> And yes I'm all for merging changes as long as it's incrementally better than trunk 👍

👌

**Comment by ciampo** [comment]
Hey @Rishit30G , @mirka and I think that the improvements in this PR are still worth pursuing, despite recognising that we still won't be able to fully support `color-mix()` syntaxes.

Here's a list of next steps for this PR:

- Apply suggestions from https://github.com/WordPress/gutenberg/pull/64224#discussion_r1718254842
- Apply similar changes to the `normalizeColorValue` (see suggested diff in https://github.com/WordPress/gutenberg/pull/64224#discussion_r1718271911)
- Consider extracting a reusing a `isSimpleCSSColor` utility that checks for CSS variable and `color-mix()`
- Add unit tests if possible

Does it sound good ?

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [reply packages/components/src/color-palette/utils.ts:1]: > We should ideally find a browser-based way (either through CSS or native JS) that will guarantee us that whatever CSS color was passed to the component, we're able to parse it, show it, and render a working color picker.

Actually, hold up my `canvas`: https://codepen.io/ciampo/pen/QWXamNe

| Screenshot 2024-08-16 at 15 29 50 | Screenshot 2024-08-16 at 15 27 28 |
|---|---|

This approach could have potential. We'd need to make sure it can't be somehow blocked by the browser, and that there are no severe performance implications (if needed, we can explore using `OffscreenCanvas` in a separate worker thread).

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [inline packages/components/src/color-palette/utils.ts]: We can simplify the code by inlining the `isSimpleCSSColor` call:

```js
	if ( isSimpleCSSColor( value ) || element === null ) {
```

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [inline packages/components/src/color-palette/utils.ts]: We should reuse the `isSimpleCSSColor` function here, something like this should work:

```diff
diff --git a/packages/components/src/color-palette/utils.ts b/packages/components/src/color-palette/utils.ts
index d5f81f3304..88e5c89dd9 100644
--- a/packages/components/src/color-palette/utils.ts
+++ b/packages/components/src/color-palette/utils.ts
@@ -17,6 +17,19 @@ import type { ColorObject, ColorPaletteProps, PaletteObject } from './types';
 
 extend( [ namesPlugin, a11yPlugin ] );
 
+/**
+ * Checks if a color value is a simple CSS color, its value doesn't need
+ * further processing to be used as a color.
+ *
+ * @param value The color value to check.
+ * @return A boolean indicating whether the color value is a simple CSS color.
+ */
+const isSimpleCSSColor = ( value: string | undefined ) => {
+	const valueIsCssVariable = /var\(/.test( value ?? '' );
+	const valueIsColorMix = /color-mix\(/.test( value ?? '' );
+	return ! valueIsCssVariable && ! valueIsColorMix;
+};
+
 export const extractColorNameFromCurrentValue = (
 	currentValue?: ColorPaletteProps[ 'value' ],
 	colors: ColorPaletteProps[ 'colors' ] = [],
@@ -26,12 +39,10 @@ export const extractColorNameFromCurrentValue = (
 		return '';
 	}
 
-	const currentValueIsCssVariable = /^var\(/.test( currentValue );
-	const currentValueIsColorMix = /color-mix\(/.test( currentValue );
-	const normalizedCurrentValue =
-		currentValueIsCssVariable || currentValueIsColorMix
-			? currentValue
-			: colord( currentValue ).toHex();
+	const currentValueIsSimpleColor = isSimpleCSSColor( currentValue );
+	const normalizedCurrentValue = currentValueIsSimpleColor
+		? colord( currentValue ).toHex()
+		: currentValue;
 
 	// Normalize format of `colors` to simplify the following loop
 	type normalizedPaletteObject = { colors: ColorObject[] };
@@ -40,10 +51,9 @@ export const extractColorNameFromCurrentValue = (
 		: [ { colors: colors as ColorObject[] } ];
 	for ( const { colors: paletteColors } of colorPalettes ) {
 		for ( const { name: colorName, color: colorValue } of paletteColors ) {
-			const normalizedColorValue =
-				currentValueIsCssVariable || currentValueIsColorMix
-					? colorValue
-					: colord( colorValue ).toHex();
+			const normalizedColorValue = currentValueIsSimpleColor
+				? colord( colorValue ).toHex()
+				: colorValue;
 
 			if ( normalizedCurrentValue === normalizedColorValue ) {
 				return colorName;
@@ -71,18 +81,6 @@ export const isMultiplePaletteArray = (
 	);
 };
 
-/**
- * Checks if a color value is a simple CSS color.
- *
- * @param value The color value to check.
- * @return A boolean indicating whether the color value is a simple CSS color.
- */
-const isSimpleCSSColor = ( value: string | undefined ): boolean => {
-	const valueIsCssVariable = /var\(/.test( value ?? '' );
-	const valueIsColorMix = /color-mix\(/.test( value ?? '' );
-	return ! valueIsCssVariable && ! valueIsColorMix;
-};
-
 /**
  * Transform a CSS variable used as background color into the color value itself.
  *

```

**Review by ciampo — COMMENTED** [review]
Thank you! 

I left some extra feedback on the current changes.

For the unit tests, we could add the `color-mix` tests to [this file](https://github.com/WordPress/gutenberg/blob/1d6368a21cc55ec8db062a1faf8715cbe7310cd2/packages/components/src/color-palette/test/utils.ts). Let me know if you need help in writing those tests.

File: packages/components/src/color-palette/utils.ts
Comment (tyxla) [inline packages/components/src/color-palette/utils.ts]: If we accept `undefined` values, we can (and IMHO should) immediately return early, without the need for the regex tests and the condition below.

File: packages/components/src/color-palette/utils.ts
Comment (tyxla) [inline packages/components/src/color-palette/utils.ts]: So if I pass `undefined`, it returns that it's a simple CSS color? Seems incorrect to me.

Comment (ciampo) [reply packages/components/src/color-palette/utils.ts]: That's a good point, the cleanest change would probably be to handle the fallback to an empty string outside of `isSimpleCSSColor`.

File: packages/components/src/color-palette/utils.ts
Comment (ciampo) [reply packages/components/src/color-palette/utils.ts]: https://github.com/WordPress/gutenberg/pull/64224#discussion_r1721496744

File: packages/components/src/color-palette/utils.ts
Comment (tyxla) [reply packages/components/src/color-palette/utils.ts]: Yep, makes sense, and we already do handle `null` outside.

**Review by ciampo — COMMENTED** [review]
I actually looked into writing some unit tests, but it looks like they would be a bit pointless since JSDom has very poor support for CSS custom properties and CSS color syntax — so our best bet would be to use a real browser environment to test this component, and that's not in the scope of this PR.

@Rishit30G , the last thing left before merging would be to add a CHANGELOG entry in [this file](https://github.com/WordPress/gutenberg/blob/eea8d0844df950103b88c0fe6a905ee1e0c16f7d/packages/components/CHANGELOG.md). I would probably add it under the `Enhancements` category.

Thank you!

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `ColorPalette`: partial support of `color-mix()` CSS colors ([#64224](https://github.com/WordPress/gutenberg/pull/64224)).
```

**Review by ciampo — APPROVED** [review]
LGTM 🚀

File: packages/components/src/color-palette/utils.ts
Comment (ramonjd) [inline packages/components/src/color-palette/utils.ts:95]: I believe this change has created a regression.

It's treating `undefined` values as false positives = non-simple values, which means it then returns `defaultView?.getComputedStyle`

### With this changes

The default alpha value is `0`, which means, by default, any custom color cannot be seen in the canvas.

https://github.com/user-attachments/assets/fb482ad7-f900-41f4-ac28-9b36915f81f7


### With these changes reverted

The alpha is `100` so, by default, my custom color changes are visible on the canvas.

https://github.com/user-attachments/assets/3f4e9b8e-a65e-4536-a34d-664423d9abbd

Comment (ramonjd) [reply packages/components/src/color-palette/utils.ts:95]: I have a fix here:

- https://github.com/WordPress/gutenberg/pull/64969

Would be good to confirm that it doesn't affect what this PR intends to do. Thanks!

**Comment by ciampo** [comment]
@fabiankaegy I don't think this PR deserves a Dev Note, since it's mostly a best-effort bug fix that prevents the component from crashing. The `color-mix` syntax is not really supported, and as such I don't think we should advertise this to WordPress developers.

I removed it from the list int he 6.7 dev notes tracking issue.

**Pull request merged**
